x-pack/apm-server/sampling/pubsub/pubsub.go

```diff
@@ -86,114 +93,163 @@ func (p *Pubsub) SubscribeSampledTraceIDs(ctx context.Context, traceIDs chan<- s
 	ticker := time.NewTicker(p.config.SearchInterval)
 	defer ticker.Stop()
 
-	// NOTE(axw) we should use the Changes API when it is implemented:
-	// https://github.com/elastic/elasticsearch/issues/1242
-
-	var lastSeqNo int64 = 0
-	var lastPrimaryTerm int64 = -1
+	observedSeqnos := make(map[string]int64)
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		case <-ticker.C:
 		}
-		for {
-			// Keep searching until there are no more new trace IDs.
-			n, err := p.searchTraceIDs(ctx, traceIDs, &lastSeqNo, &lastPrimaryTerm)
+		if err := p.searchTraceIDs(ctx, traceIDs, observedSeqnos); err != nil {
+			// Errors may occur due to rate limiting, or while the index is
+			// still being created, so just log and continue.
+			p.config.Logger.With(logp.Error(err)).Debug("error searching for trace IDs")
+		}
+	}
+}
+
+// searchTraceIDs searches the configured data stream for new sampled trace IDs, sending them to the out channel.
+//
+// searchTraceIDs works by fetching the global checkpoint for each index backing the data stream, and comparing
+// this to the most recently observed sequence number for the indices. If the global checkpoint is greater, then
+// we search through every document with a sequence number greater than the most recently observed, and less than
+// or equal to the global checkpoint.
+//
+// Searches may not return everything up to the global checkpoint, depending on whether the index has been
+// refreshed since the global checkpoint was last updated. As such, we will only progress the locally observed
```

Comment:
This still risks not seeing documents. Unfortunately, you have to refresh currently to avoid that. The global checkpoints returned are independent of refresh. The refreshed state represents and old set of documents. Let us assume you find global checkpoint 10. The state could then be:

unrefreshed: 0,1,2,3,4,5,6,7,8,9,10,12,13
refreshed: 0,1,2,3,4,5,6,7,9,10,12

So while you do not consider 12 (since it is above GCP), you did loose document with seqno 8 by only looking at refreshed state.

There are 2 solutions:

1. Do a refresh between the stats call and the search call.
2. Only consider docs with consecutive seq-nos

The problem with 1 is that performance will degrade if you do too many refreshes.

The problem with 2 is that actually seeing the docs is delayed until a poll round after a refresh (something have to refresh, could be just the normal refresh timer).

The best choice depends on the polling rate. If you potentially poll multiple times per second, the second options is probably better. If you poll only every 10 seconds or so, I think doing an explicit refresh is probably best.

Reply:
Thanks - I'll add an explicit refresh in then. We'll query every 30 seconds by default. I was trying to minimise the load on Elasticsearch, as there will potentially be many APM Servers. Hopefully this is still a temporary measure anyway, and that we can move over to the global checkpoints API eventually.